cmd/erasure-server-pool-decom.go

```diff
@@ -0,0 +1,881 @@
+	if _, err = p.UnmarshalMsg(data[4:]); err != nil {
+		return false, err
+	}
+
+	switch p.Version {
+	case poolMetaVersionV1:
+	default:
+		return false, fmt.Errorf("unexpected pool meta version: %d", p.Version)
+	}
+
+	type poolInfo struct {
+		position  int
+		completed bool
+	}
+
+	rememberedPools := make(map[string]poolInfo)
+	for idx, pool := range p.Pools {
+		complete := false
+		if pool.Decommission != nil && pool.Decommission.Complete {
+			complete = true
+		}
+		rememberedPools[pool.CmdLine] = poolInfo{
+			position:  idx,
+			completed: complete,
+		}
+	}
+
+	specifiedPools := make(map[string]int)
+	for idx, pool := range pools {
+		specifiedPools[pool.endpoints.CmdLine] = idx
+	}
+
+	// Check if specified pools need to remove decommissioned pool.
+	for k := range specifiedPools {
+		pi, ok := rememberedPools[k]
+		if ok && pi.completed {
+			return false, fmt.Errorf("pool(%s) = %s is decommissioned, please remove from server command line", humanize.Ordinal(pi.position+1), k)
+		}
+	}
+
+	// check if remembered pools are in right position or missing from command line.
+	for k, pi := range rememberedPools {
+		if pi.completed {
+			continue
+		}
+		_, ok := specifiedPools[k]
+		if !ok {
+			return false, fmt.Errorf("pool(%s) = %s is not specified, please specify on server command line", humanize.Ordinal(pi.position+1), k)
+		}
+	}
+
+	// check when remembered pools and specified pools are same they are at the expected position
+	if len(rememberedPools) == len(specifiedPools) {
+		for k, pi := range rememberedPools {
+			pos, ok := specifiedPools[k]
+			if !ok {
+				return false, fmt.Errorf("pool(%s) = %s is not specified, please specify on server command line", humanize.Ordinal(pi.position+1), k)
+			}
+			if pos != pi.position {
+				return false, fmt.Errorf("pool order change detected for %s, expected position is (%s) but found (%s)", k, humanize.Ordinal(pi.position+1), humanize.Ordinal(pos+1))
+			}
+		}
+	}
+
+	return len(rememberedPools) != len(specifiedPools), nil
+}
+
+func (p *poolMeta) CountItem(idx int, size int64, failed bool) {
+	pd := p.Pools[idx].Decommission
+	if pd != nil {
+		if failed {
+			pd.ItemsDecommissionFailed++
+			pd.BytesFailed += uint64(size)
+		} else {
+			pd.ItemsDecommissioned++
+			pd.BytesDone += uint64(size)
+		}
+		p.Pools[idx].Decommission = pd
+	}
+}
+
+func (p *poolMeta) updateAfter(ctx context.Context, idx int, pools []*erasureSets, duration time.Duration) error {
+	if p.Pools[idx].Decommission == nil {
+		return errInvalidArgument
+	}
+	if time.Since(p.Pools[idx].LastUpdate) > duration {
+		p.Pools[idx].LastUpdate = time.Now().UTC()
+		return p.save(ctx, pools)
+	}
+	return nil
+}
+
+func (p poolMeta) save(ctx context.Context, pools []*erasureSets) error {
+	data := make([]byte, 4, p.Msgsize()+4)
+
+	// Initialize the header.
+	binary.LittleEndian.PutUint16(data[0:2], poolMetaFormat)
+	binary.LittleEndian.PutUint16(data[2:4], poolMetaVersion)
+
+	buf, err := p.MarshalMsg(data)
```

Comment:
Shouldn't this be,
```suggestion
	buf, err := p.MarshalMsg(data[4:])
```

Reply:
No message pack returns the `data` first 4 bytes without interpretation @krisis - it uses the bytes that are not used automatically. 

The returned `buf` has all the bytes needed included the first 4 bytes.

Reply:
check the implementation in `cmd/tier.go`